openedx/core/djangoapps/schedules/tasks.py

```diff
@@ -309,8 +314,98 @@ def _upgrade_reminder_schedules_for_bin(site, target_day, bin_num, org_list, exc
+def course_update_schedule_bin(
+    site_id, target_day_str, day_offset, bin_num, org_list, exclude_orgs=False, override_recipient_email=None,
+):
+    target_day = deserialize(target_day_str)
+    msg_type = CourseUpdate()
+
+    for (user, language, context) in _course_update_schedules_for_bin(
+        Site.objects.get(id=site_id),
+        target_day,
+        day_offset,
+        bin_num,
+        org_list,
+        exclude_orgs
+    ):
+        msg = msg_type.personalize(
+            Recipient(
+                user.username,
+                override_recipient_email or user.email,
+            ),
+            language,
+            context,
+        )
+        _course_update_schedule_send.apply_async((site_id, str(msg)), retry=False)
+
+
+@task(ignore_result=True, routing_key=ROUTING_KEY)
+def _course_update_schedule_send(site_id, msg_str):
+    site = Site.objects.get(pk=site_id)
+    if not ScheduleConfig.current(site).deliver_course_update:
+        return
+
+    msg = Message.from_string(msg_str)
+    ace.send(msg)
+
+
+def _course_update_schedules_for_bin(site, target_day, day_offset, bin_num, org_list, exclude_orgs=False):
+    week_num = abs(day_offset) / 7
+    beginning_of_day = target_day.replace(hour=0, minute=0, second=0)
+    schedules = get_schedules_with_target_date_by_bin_and_orgs(
+        schedule_date_field='start',
+        target_date=beginning_of_day,
+        bin_num=bin_num,
+        num_bins=COURSE_UPDATE_NUM_BINS,
+        org_list=org_list,
+        exclude_orgs=exclude_orgs,
+        order_by='enrollment__course',
+    )
+
+    LOG.debug('Course Update: Query = %r', schedules.query.sql_with_params())
+
+    for schedule in schedules:
+        enrollment = schedule.enrollment
+        try:
+            week_summary = get_course_week_summary(enrollment.course_id, week_num)
+        except CourseUpdateDoesNotExist:
+            continue
+
+        user = enrollment.user
+        course_id_str = str(enrollment.course_id)
+
+        template_context = get_base_template_context(site)
+        template_context.update({
+            'student_name': user.profile.name,
+            'user_personal_address': user.profile.name if user.profile.name else user.username,
+            'course_name': schedule.enrollment.course.display_name,
+            'course_url': absolute_url(site, reverse('course_root', args=[str(schedule.enrollment.course_id)])),
+            'week_num': week_num,
+            'week_summary': week_summary,
+
+            # This is used by the bulk email optout policy
+            'course_ids': [course_id_str],
+        })
+
+        yield (user, schedule.enrollment.course.language, template_context)
+
+
+@request_cached
+def get_course_week_summary(course_id, week_num):
+    if COURSE_UPDATE_WAFFLE_FLAG.is_enabled(course_id):
+        course = modulestore().get_course(course_id)
+        return course.week_summary(week_num)
+    else:
+        raise CourseUpdateDoesNotExist()
```

Comment:
Where do you fall on the `None` vs exceptions question for control flow? I probably would have done this using `None`, because it seems expected that we might not have an update for a given week.

Reply:
yeah, i would have normally also used `None` here, but i was experimenting with exceptions, given our discussion earlier.  I'm thinking of giving it a try (pun-intended) to see how it will play out as we evolve this code.  what do you think?

Reply:
Also, see my comment above https://github.com/edx/edx-platform/pull/16219#discussion_r144656272.